File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>

using namespace std;

vector<int>v[201];
int visited[201];

void dfs(int x){
    if(visited[x]==1) return;
    visited[x]=1;
    for(int i=0;i<v[x].size();i++){
        int k =v[x][i];
        dfs(k);
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    for(int i=0;i<computers.size();i++){
        for(int j=0;j<computers[0].size();j++){
            if(i!=j) {
                if(computers[i][j]==1){
                    v[i].push_back(j);
                }
            }
        }
    }
    
    for(int i=0;i<n;i++){
        if(visited[i]==0) {
            answer++;
            dfs(i);
        }
    }
    
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    
    private static List<Integer>[] adjList;
    private static boolean[] visited;
    
    private static void dfs(int x){
        if(visited[x]) return;
        visited[x] = true;
        
        for(int i=0;i<adjList[x].size();i++){
            dfs(adjList[x].get(i));
        }
    }
    
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        adjList = new ArrayList[n];
        visited = new boolean[n];
        
        for(int i=0;i<n;i++){
            adjList[i] = new ArrayList<>();
        }
        
        for(int i=0;i<n-1;i++){
            for(int j=i+1;j<n;j++){
                if(computers[i][j]==1){
                    adjList[i].add(j);
                    adjList[j].add(i);
                }
            }
        }
        
        for(int i=0;i<n;i++){
            if(!visited[i]){
                answer++;
                dfs(i);
            }
        }
        
        
        
        
        return answer;
    }
}